File: php/src/Algorithms.php
<?php

declare(strict_types=1);

namespace Wizardlabz\Kiahk;

/**
 * Calendar conversion primitives for the Coptic (Alexandrian) calendar.
 * Identical results to the JS / Python / Go / Dart / Swift / C# / C ports
 * against core/test-vectors.json. See core/algorithms.md for the spec.
 */
final class Algorithms
{
    /** JDN of 1 Tout, year 1 AM (Coptic epoch). */
    public const COPTIC_EPOCH = 1825030;

    private function __construct() {}

    /** Gregorian date → Julian Day Number (Fliegel & Van Flandern). */
    public static function gregorianToJdn(int $year, int $month, int $day): int
    {
        $a = intdiv(14 - $month, 12);
        $y = $year + 4800 - $a;
        $m = $month + 12 * $a - 3;
        return $day
            + intdiv(153 * $m + 2, 5)
            + 365 * $y
            + intdiv($y, 4) - intdiv($y, 100) + intdiv($y, 400)
            - 32045;
    }

    /**
     * Julian Day Number → Gregorian date.
     * @return array{0:int,1:int,2:int} [year, month, day]
     */
    public static function jdnToGregorian(int $jdn): array
    {
        $a = $jdn + 32044;
        $b = intdiv(4 * $a + 3, 146097);
        $c = $a - intdiv(146097 * $b, 4);
        $d = intdiv(4 * $c + 3, 1461);
        $e = $c - intdiv(1461 * $d, 4);
        $m = intdiv(5 * $e + 2, 153);
        $day = $e - intdiv(153 * $m + 2, 5) + 1;
        $month = $m + 3 - 12 * intdiv($m, 10);
        $year = 100 * $b + $d - 4800 + intdiv($m, 10);
        return [$year, $month, $day];
    }

    /** Coptic date → Julian Day Number. */
    public static function copticToJdn(int $cYear, int $cMonth, int $cDay): int
    {
        return self::COPTIC_EPOCH - 1
            + 365 * ($cYear - 1)
            + intdiv($cYear, 4)
            + 30 * ($cMonth - 1)
            + $cDay;
    }

    /**
     * Julian Day Number → Coptic date.
     * @return array{0:int,1:int,2:int} [cYear, cMonth, cDay]
     */
    public static function jdnToCoptic(int $jdn): array
    {
        $r = $jdn - self::COPTIC_EPOCH;
        $cYear = intdiv(4 * $r + 1463, 1461);
        $dayOfYr = $r - 365 * ($cYear - 1) - intdiv($cYear, 4);
        $cMonth = intdiv($dayOfYr, 30) + 1;
        $cDay = $dayOfYr - 30 * ($cMonth - 1) + 1;
        return [$cYear, $cMonth, $cDay];
    }

    /**
     * Coptic Easter (Pascha) for a Gregorian year.
     * Julian computus (Meeus) + +13 Julian→Gregorian shift, valid 1900-03-01..2100-02-28.
     * @return array{0:int,1:int,2:int} [year, month, day] (Gregorian)
     */
    public static function computeEaster(int $gregorianYear): array
    {
        $a = $gregorianYear % 4;
        $b = $gregorianYear % 7;
        $c = $gregorianYear % 19;
        $d = (19 * $c + 15) % 30;
        $e = (2 * $a + 4 * $b - $d + 34) % 7;
        // PHP's % preserves sign of dividend, but $e is always >= 0 here because
        // the +34 ensures the numerator is positive for all gregorianYear >= 0.
        $f = intdiv($d + $e + 114, 31);
        $g = (($d + $e + 114) % 31) + 1;
        $jdn = self::gregorianToJdn($gregorianYear, $f, $g) + 13;
        return self::jdnToGregorian($jdn);
    }

    /**
     * Add `days` to a Gregorian date.
     * @return array{0:int,1:int,2:int} [year, month, day]
     */
    public static function addDays(int $year, int $month, int $day, int $days): array
    {
        return self::jdnToGregorian(self::gregorianToJdn($year, $month, $day) + $days);
    }
}

File: c/src/algorithms.c
#include "kiahk.h"

/* JDN of 1 Tout, year 1 AM (Coptic epoch). */
#define KIAHK_COPTIC_EPOCH 1825030

int kiahk_gregorian_to_jdn(int year, int month, int day) {
    int a = (14 - month) / 12;
    int y = year + 4800 - a;
    int m = month + 12 * a - 3;
    return day
        + (153 * m + 2) / 5
        + 365 * y
        + y / 4
        - y / 100
        + y / 400
        - 32045;
}

void kiahk_jdn_to_gregorian(int jdn, int *out_year, int *out_month, int *out_day) {
    int a = jdn + 32044;
    int b = (4 * a + 3) / 146097;
    int c = a - (146097 * b) / 4;
    int d = (4 * c + 3) / 1461;
    int e = c - (1461 * d) / 4;
    int m = (5 * e + 2) / 153;
    *out_day = e - (153 * m + 2) / 5 + 1;
    *out_month = m + 3 - 12 * (m / 10);
    *out_year = 100 * b + d - 4800 + m / 10;
}

int kiahk_coptic_to_jdn(int year, int month, int day) {
    return KIAHK_COPTIC_EPOCH - 1
        + 365 * (year - 1)
        + year / 4
        + 30 * (month - 1)
        + day;
}

void kiahk_jdn_to_coptic(int jdn, int *out_year, int *out_month, int *out_day) {
    int r = jdn - KIAHK_COPTIC_EPOCH;
    int year = (4 * r + 1463) / 1461;
    int day_of_year = r - 365 * (year - 1) - year / 4; /* 0-indexed */
    int month = day_of_year / 30 + 1;
    int day = day_of_year - 30 * (month - 1) + 1;
    *out_year = year;
    *out_month = month;
    *out_day = day;
}

void kiahk_gregorian_to_coptic(int gy, int gm, int gd, int *cy, int *cm, int *cd) {
    int jdn = kiahk_gregorian_to_jdn(gy, gm, gd);
    kiahk_jdn_to_coptic(jdn, cy, cm, cd);
}

void kiahk_coptic_to_gregorian(int cy, int cm, int cd, int *gy, int *gm, int *gd) {
    int jdn = kiahk_coptic_to_jdn(cy, cm, cd);
    kiahk_jdn_to_gregorian(jdn, gy, gm, gd);
}

void kiahk_compute_easter(int gregorian_year, int *out_year, int *out_month, int *out_day) {
    int a = gregorian_year % 4;
    int b = gregorian_year % 7;
    int c = gregorian_year % 19;
    int d = (19 * c + 15) % 30;
    int e = (2 * a + 4 * b - d + 34) % 7;
    int f = (d + e + 114) / 31;
    int g = (d + e + 114) % 31 + 1;
    int jdn = kiahk_gregorian_to_jdn(gregorian_year, f, g) + 13;
    kiahk_jdn_to_gregorian(jdn, out_year, out_month, out_day);
}

void kiahk_add_days(int year, int month, int day, int days,
                    int *out_year, int *out_month, int *out_day) {
    int jdn = kiahk_gregorian_to_jdn(year, month, day) + days;
    kiahk_jdn_to_gregorian(jdn, out_year, out_month, out_day);
}
